### DISCUSSION TITLE
Websockets not working

### DISCUSSION THREAD

**Discussion by fancysnake** [outline/outline#2222] [Self Hosting] [answered]
Hi, I need some help configuring Outline.
I have a docker compose config with Outline running on port 3000 and an external nginx with this site config:

```
upstream outline {
    server localhost:3000;
}


server {
    server_name mydomain.net;
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/mydomain.net/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/mydomain.net/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

    location / {
        proxy_pass http://outline;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $host;
        proxy_redirect off;
        proxy_http_version 1.1;
    }

}

server {
    if ($host = mydomain.net) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

    server_name mydomain.net;
    listen 80;
    return 404; # managed by Certbot
}
```

Unfortunarely, websockets are not working with this setup. I got these errors in firefox:
```
Firefox can’t establish a connection to the server at wss://mydomain.net/realtime/?EIO=3&transport=websocket. websocket.js:124:12
The connection to wss://mydomain.net/realtime/?EIO=3&transport=websocket was interrupted while the page was loading. websocket.js:124:12
```

So, the site is functional, but for example, when I create a new document it's not visible until I refresh the page.
Could you help me in getting this right?

**Comment by fancysnake** [comment, marked as answer]
Nevermind, I've found solution. Needed to add these lines to nginx:
```
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
```

**Reply by gauravk-in** [reply]
Can you please help me? I have an Nginx Reverse Proxy which allows websockets and uses SSL and I have added these lines, but I still get the same error.

**Reply by inyerade** [reply]
can you post your nginx config file?

**Comment by martinwittmann** [comment]
I'm facing the same issue and setting the headers does not fix it. Documents can be edited after a minute or so, but not right after creating them.

Here is the nginx config:

```
server {
  server_name outline.example.com;
  listen 443;
  listen [::]:443;
  http2 on; 

  # ...
  # Removed configs related to gzip and ssl.
  # ...

  location / { 
    proxy_pass http://outline:3000;
    proxy_set_header Host $http_host;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header X-Real-IP $remote_addr;

    # Possible fix for outline not allowing editing of documents.
    # See https://github.com/outline/outline/discussions/3546
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }
}

```

**Reply by martinwittmann** [reply]
I found the issue in my case. I set the Upgrade header multiple times which causes the issue

**Reply by ainslied** [reply]
Hi,

I encountered the same issue, and none of the solutions provided above worked for me.

However, I found that the root cause was exactly the one described in this discussion: https://github.com/outline/outline/discussions/8000

The problem was that leaving COLLABORATION_URL empty causes WebSocket connections to fail.
To fix it, I simply set it explicitly in the environment variables:
```
URL=https://outline.myserver.net
COLLABORATION_URL=wss://outline.myserver.net
```

Hope this helps others facing the same issue.
